certbot/error_handler.py

```diff
@@ -86,19 +108,34 @@ def set_signal_handlers(self):
                 self.prev_handlers[signum] = prev_handler
                 signal.signal(signum, self._signal_handler)
 
-    def reset_signal_handlers(self):
+    def _reset_signal_handlers(self):
         """Resets signal handlers for signals in _SIGNALS."""
         for signum in self.prev_handlers:
             signal.signal(signum, self.prev_handlers[signum])
+        out = dict((k, v) for k, v in self.prev_handlers.items())
         self.prev_handlers.clear()
+        return out
 
     def _signal_handler(self, signum, unused_frame):
-        """Calls registered functions and the previous signal handler.
+        """Replacement function for handling recieved signals.
+
+        Store the recieved signal. If we are executing the code block in
+        the body of the context manager, stop by raising signal exit.
 
         :param int signum: number of current signal
 
         """
-        logger.debug("Singal %s encountered", signum)
-        self.call_registered()
-        signal.signal(signum, self.prev_handlers[signum])
-        os.kill(os.getpid(), signum)
+        self.received_signals.append(signum)
+        if not self.body_executed:
+            raise errors.SignalExit
+
+    def _call_signals(self, prev_handlers):
+        """Finally call the deferred signals.
+
+        :param int signum: signal number
+
+        """
+        for signum in self.received_signals:
+            logger.debug("Calling signal %s", signum)
+            signal.signal(signum, prev_handlers[signum])
```

Comment:
Because we're calling `_reset_signal_handlers` first, we don't need to do this now. The call to `signal.signal` in `_reset_signal_handlers` already restored the previous signal handlers.